crypto/x509/x509_vfy.c

```diff
@@ -741,7 +742,17 @@ static int check_id(X509_STORE_CTX *ctx)
         if (!check_id_error(ctx, X509_V_ERR_IP_ADDRESS_MISMATCH))
             return 0;
     }
-    return 1;
+
+    /*
+     * When verifying SSL server certificates, require that an
+     * identity be set for validating the certificate subject unless
+     * the X509_V_FLAG_ALLOW_NO_SUBJECT_CHECK flags is set instead.
+     */
+    if (vpm->purpose != X509_PURPOSE_SSL_SERVER
```

Comment:
Instead of `vpm->purpose != X509_PURPOSE_SSL_SERVER` it would have to be `vpm->purpose == X509_PURPOSE_SSL_CLIENT`.  Otherwise, this breaks `X509_PURPOSE_SMIME_SIGN` and possibly others.

Reply:
When we are a client, the purpose of the certificate we are checking is "server".

I learned this the fun way, of course...